src/ledger.ts

```diff
@@ -1280,3 +1287,183 @@ export class LedgerSignMessage extends LedgerBitcoinInteraction {
     });
   }
 }
+
+interface RegistrationConstructorArguments {
+  name: string;
+  braid: Braid;
+}
+
+/**
+ * A base class for any interactions that need to interact with a registered wallet
+ * by providing a base constructor that will generate the key origins and the policy
+ * from a given braid as well as methods for registering and returning a policy hmac
+ */
+export class LedgerBitcoinV2WithRegistrationInteraction extends LedgerBitcoinInteraction {
+  walletPolicy: MutlisigWalletPolicy;
+  policyHmac?: Buffer;
+  policyId?: Buffer;
+
+  isLegacySupported = false;
+  isV2Supported = true;
+
+  constructor({ name, braid }: RegistrationConstructorArguments) {
+    super();
+    const keyOrigins = getKeyOriginsFromBraid(braid);
+    const template = getPolicyTemplateFromBraid(braid);
+    this.walletPolicy = new MutlisigWalletPolicy({
+      name,
+      keyOrigins,
+      template,
+    });
+  }
+
+  messages() {
+    const messages = super.messages();
+
+    if (!this.policyHmac) {
+      messages.push({
+        state: ACTIVE,
+        level: INFO,
+        code: "ledger.confirm.address",
+        version: ">=2.1.0",
+        text: `Your Ledger will ask you to register your wallet info first. This allows the device to derive an address for a verified quorum.`,
+        action: LEDGER_RIGHT_BUTTON,
+      });
+    }
+
+    return messages;
+  }
+
+  registerWallet(): Promise<Buffer> {
+    if (this.policyHmac) return Promise.resolve(this.policyHmac);
+    else {
+      return this.withApp(async (app: AppClient) => {
+        const policy = this.walletPolicy.toLedgerPolicy();
+        const [policyId, policyHmac] = await app.registerWallet(policy);
+        this.policyHmac = policyHmac;
+        this.policyId = policyId;
+        policyHmac;
+      });
+    }
+  }
+
+  async run(): Promise<unknown> {
+    await super.run();
+    return await this.registerWallet();
+  }
+}
+
+interface ConfirmAddressConstructorArguments
+  extends RegistrationConstructorArguments {
+  // the expected address to compare against
+  expected: string;
+  // whether or not to display the address to the user
+  display: boolean;
+  // the index
+  addressIndex: number;
+}
+
+/**
+ * Interaction for confirming an address on a ledger device. Requires a registered
+ * wallet to complete successfully. Only supported on Ledger v2.1.0 or above.
+ */
+export class LedgerConfirmMultisigAddress extends LedgerBitcoinV2WithRegistrationInteraction {
+  braidIndex: 0 | 1;
+  addressIndex: number;
+  expected?: string;
+
+  display = true;
+
+  constructor({
+    braid,
+    name,
+    addressIndex,
+    display,
+    expected,
+  }: ConfirmAddressConstructorArguments) {
+    super({ braid, name });
+
+    const braidIndex = Number(braid.index);
+    if (braidIndex !== 1 && braidIndex !== 0)
+      throw new Error(`Invalid braid index ${braidIndex}`);
+    this.braidIndex = braidIndex;
+
+    const index = Number(addressIndex);
```

Comment:
Should already be a number since it's a number in the arguments

Reply:
I was unsure how this would be dealt with in non-TS environments so I left it there to just be really really sure.